File: src/main/java/com/ryderbelserion/discord/api/commands/CommandEngine.java
package com.ryderbelserion.discord.api.commands;

import net.dv8tion.jda.api.Permission;
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
import net.dv8tion.jda.api.hooks.ListenerAdapter;
import org.jetbrains.annotations.NotNull;

public abstract class CommandEngine extends ListenerAdapter {

    private final String name;
    private final String description;
    private final Permission permission;

    public CommandEngine(@NotNull final String name,
                         @NotNull final String description,
                         @NotNull final Permission permission) {
        this.name = name;
        this.description = description;
        this.permission = permission;
    }

    protected abstract void perform(@NotNull final CommandContext context);

    @Override
    public void onSlashCommandInteraction(@NotNull SlashCommandInteractionEvent event) {
        final CommandContext context = new CommandContext(event);

        if (!event.getName().equalsIgnoreCase(this.name)) return;

        perform(context);
    }

    public @NotNull final String getName() {
        return this.name;
    }

    public @NotNull final String getDescription() {
        return this.description;
    }

    public @NotNull final Permission getPermission() {
        return this.permission;
    }
}

File: src/main/kotlin/com/ryderbelserion/common/command/CommandEngine.kt
package com.ryderbelserion.common.command

import net.dv8tion.jda.api.Permission
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.hooks.ListenerAdapter

abstract class CommandEngine(val name: String, val desc: String, val permission: Permission) : ListenerAdapter() {

    abstract fun perform(context: CommandContext)

    open fun execute(context: CommandContext) {
        perform(context)
    }

    override fun onSlashCommandInteraction(event: SlashCommandInteractionEvent) {
        val context = CommandContext(event)

        if (event.name != name) return

        execute(context)
    }
}